src/jvm/main/org/jetbrains/kotlinx/lincheck/ObjectTraverser.kt

```diff
@@ -39,107 +35,81 @@ import kotlin.coroutines.Continuation
-                    if (shouldAnalyseObjectRecursively(value, objectNumberMap)) {
-                        if (value != null) {
-                            enumerateObjects(value, processedObjects, objectNumberMap)
-                        }
-                    }
+                if (isAtomicFU(jumpObj)) {
+                    jumpObj =
+                        if (jumpObj is kotlinx.atomicfu.AtomicBoolean) readNextJumpObjectByFieldName("_value")
+                        else readNextJumpObjectByFieldName("value")
                 }
-            } catch (e: Throwable) {
-                e.printStackTrace()
-                // Ignore
             }
-        }
-        clazz = clazz.superclass
-    }
-}
 
-private fun readField(obj: Any?, field: Field): Any? {
-    if (!field.type.isPrimitive) {
-        return readFieldViaUnsafe(obj, field, Unsafe::getObject)
-    }
-    return when (field.type) {
-        Boolean::class.javaPrimitiveType    -> readFieldViaUnsafe(obj, field, Unsafe::getBoolean)
-        Byte::class.javaPrimitiveType       -> readFieldViaUnsafe(obj, field, Unsafe::getByte)
-        Char::class.javaPrimitiveType       -> readFieldViaUnsafe(obj, field, Unsafe::getChar)
-        Short::class.javaPrimitiveType      -> readFieldViaUnsafe(obj, field, Unsafe::getShort)
-        Int::class.javaPrimitiveType        -> readFieldViaUnsafe(obj, field, Unsafe::getInt)
-        Long::class.javaPrimitiveType       -> readFieldViaUnsafe(obj, field, Unsafe::getLong)
-        Double::class.javaPrimitiveType     -> readFieldViaUnsafe(obj, field, Unsafe::getDouble)
-        Float::class.javaPrimitiveType      -> readFieldViaUnsafe(obj, field, Unsafe::getFloat)
-        else                                -> error("No more types expected")
+            if (jumpObj != null) {
+                objectNumberMap[jumpObj] = getObjectNumber(jumpObj.javaClass, jumpObj)
+                if (shouldAnalyseObjectRecursively(jumpObj, objectNumberMap)) jumpObj else null
+            }
+            else null
+        }
     }
-}
 
-private fun isAtomic(value: Any?): Boolean {
-    if (value == null) return false
-    return value.javaClass.canonicalName.let {
-        it == "java.util.concurrent.atomic.AtomicInteger" ||
-        it == "java.util.concurrent.atomic.AtomicLong" ||
-        it == "java.util.concurrent.atomic.AtomicReference" ||
-        it == "java.util.concurrent.atomic.AtomicBoolean"
-    }
+    traverseObjectGraph(
+        obj,
+        onArrayElement = { _, _, value ->
+            if (value?.javaClass?.isEnum == true) null
+            else try {
+                processObject(value)
+            } catch (e: Throwable) {
+                e.printStackTrace()
+                null
+            }
+        },
+        onField = { _, f, value ->
+            if (
+                Modifier.isStatic(f.modifiers) ||
+                f.isEnumConstant ||
+                f.name == "serialVersionUID"
+            ) null
+            else try {
```

Comment:
Please make it look like the code below, with `{ .. }` brackets. Please also check other places. 

```
if (...) {
 ...
} else {
  ...
}
```